packages/salesforcedx-vscode-apex/src/embeddedSoql/soqlCompletion.ts

```diff
@@ -0,0 +1,145 @@
+  );
+  return soqlItem
+    ? { queryText: soqlItem.detail as string, location: soqlItem.data }
+    : undefined;
+}
+function insideApexBindingExpression(
+  document: TextDocument,
+  soqlQuery: string,
+  position: Position
+): boolean {
+  // Simple heuristic to detect when cursor is on a binding expression
+  // (which might have been missed by Apex LSP)
+  const rangeAtCursor = document.getWordRangeAtPosition(
+    position,
+    /[:(_\.\w)]+/
+  );
+  const wordAtCursor = rangeAtCursor
+    ? document.getText(rangeAtCursor)
+    : undefined;
+
+  return !!wordAtCursor && wordAtCursor.startsWith(':');
+}
+
+function getSOQLVirtualContent(
+  document: TextDocument,
+  position: Position,
+  soqlBlock: { queryText: string; location: any }
+): string {
+  const eol = eolForDocument(document);
+  let content = document
+    .getText()
+    .split(eol)
+    .map(line => {
+      return ' '.repeat(line.length);
+    })
+    .join(eol);
+
+  content =
+    content.slice(0, soqlBlock.location.startIndex) +
+    soqlBlock.queryText +
+    content.slice(soqlBlock.location.startIndex + soqlBlock.queryText.length);
+
+  return content;
+}
+
+export const soqlMiddleware: Middleware = {
+  provideCompletionItem: async (document, position, context, token, next) => {
+    const apexCompletionItems = await next(document, position, context, token);
+    if (!apexCompletionItems) {
+      return;
+    }
+
+    const items: ProtocolCompletionItem[] = Array.isArray(apexCompletionItems)
+      ? (apexCompletionItems as ProtocolCompletionItem[])
+      : ((apexCompletionItems as CompletionList)
+          .items as ProtocolCompletionItem[]);
+
+    const soqlBlock = insideSOQLBlock(items);
+    if (soqlBlock) {
+      if (
+        !insideApexBindingExpression(document, soqlBlock.queryText, position)
+      ) {
+        return await doSOQLCompletion(document, position, context, soqlBlock);
+      } else {
+        return items.filter(
+          i => i.label !== SOQL_SPECIAL_COMPLETION_ITEM_LABEL
+        );
+      }
+    } else return apexCompletionItems;
+  }
+};
+
+async function doSOQLCompletion(
+  document: TextDocument,
+  position: Position,
+  context: any,
+  soqlBlock: any
+): Promise<CompletionItem[] | CompletionList<CompletionItem>> {
+  const originalUri = document.uri.toString();
+  virtualDocumentContents.set(
+    originalUri,
+    getSOQLVirtualContent(document, position, soqlBlock)
+  );
+
+  const vdocUriString = `embedded-soql://soql/${encodeURIComponent(
+    originalUri
+  )}.soql`;
+  const vdocUri = Uri.parse(vdocUriString);
+  const soqlCompletions = await commands.executeCommand<CompletionList>(
+    'vscode.executeCompletionItemProvider',
+    vdocUri,
+    position,
+    // new vscode.Position(0, 0),
+    context.triggerCharacter
+  );
+  return soqlCompletions || [];
+}
+
+function eolForDocument(doc: TextDocument) {
+  switch (doc.eol) {
+    case EndOfLine.LF:
+      return '\n';
+    case EndOfLine.CRLF:
+      return '\r\n';
```

Comment:
Unless the codecov report is wrong, might want to verify this case for windows

Reply:
Both: codecov doesn't capture any of the code-completion tests. *And* I have not manually tested any of this on Windows, I'll try to find a volunteer :-)

Reply:
Cool cool. You can also probably have a test case that correctly builds a virtual document from a document with CRLF.